src/mlir/generator.cc

```diff
@@ -164,21 +171,35 @@ namespace mlir::verona
 
     // Return last value (or none)
     // TODO: Implement multiple return values for tuples
-    if (last->hasValue() && last->get().getType() != retTy)
+    auto retTy = func.getType();
+    bool hasLast = last->hasValue();
+    bool hasRetTy = retTy.getNumResults() > 0;
+    if (hasLast && hasRetTy)
     {
-      // Cast type (we trust the ast)
-      last =
-        genOperation(last->get().getLoc(), "verona.cast", {last->get()}, retTy);
+      // Function has return value and there is a last value,
+      // check types, cast if not the same, return.
+      if (last->get().getType() != retTy.getResult(0))
+      {
+        last = genOperation(
+          last->get().getLoc(),
+          "verona.cast",
+          {last->get()},
+          retTy.getResult(0));
+        if (auto err = last.takeError())
+          return std::move(err);
+      }
+      builder.create<mlir::ReturnOp>(getLocation(ast), last->getAll());
```

Comment:
Cast only if different feels like a common pattern, should this be factored out?

Reply:
It should, but not in this commit. I'm planning a bigger refactory to add more MLIR helpers, so that existing lowering functions (`parse...`) can reuse more code, and this will go into the pool. See #258.